File: Shapes/Cone/Android/app/src/main/java/com/dbooker/livreopengl/Cone.java
package com.dbooker.livreopengl;

import com.dbooker.livreopengl.glmatrix.vec3;
import com.dbooker.livreopengl.mesh.Mesh;
import com.dbooker.livreopengl.mesh.MeshModuleTopology;
import com.dbooker.livreopengl.mesh.MeshVertex;


// Définition de la classe Cone
public class Cone
{
    private Mesh m_Mesh;

    /**
     * constructeur
     * @param triangle_count : nombre de secteurs (4 pour une pyramide à base carrée, bcp plus pour un cône)
     */
    public Cone(int triangle_count) throws Exception
    {
        // créer le maillage : une grille rectangulaire
        m_Mesh = new Mesh("Cone");
        MeshModuleTopology topology = new MeshModuleTopology(m_Mesh);
        int num0 = topology.addRectangularSurface(triangle_count, 2, "cone %d-%d", true, false);

        // définir les vertices du cône
        for (int is=0; is<triangle_count; is++) {
            // vertex sur la pointe
            MeshVertex s0 = m_Mesh.getVertex(is + 0*triangle_count + num0);
            s0.setCoord(getCoord((is + 0.5f) / triangle_count, 0.0f));

            // vertex sur le bord
            MeshVertex s1 = m_Mesh.getVertex(is + 1*triangle_count + num0);
            s1.setCoord(getCoord((is + 0.0f) / triangle_count, 1.0f));
        }

        // calculer les normales
        m_Mesh.computeNormals();
    }


    /**
     * fournit le maillage de cet objet
     * @return maillage
     */
    public Mesh getMesh()
    {
        return m_Mesh;
    }


    /**
     * Cette méthode retourne les coordonnées 3D d'un point identifié par (a,b)
     * @param a : paramètre variant entre 0 et 1 en tournant autour du centre
     * @param b : rayon du point, distance à l'axe Y
     * @return vec3 contenant les coordonnées
     */
    private vec3 getCoord(float a, float b)
    {
        // angle de ce point
        float angle = (float) (a * 2.0*Math.PI);

        float x = (float) (b * Math.sin(angle));
        float y = 1.0f - b;
        float z = (float) (b * Math.cos(angle));
        return vec3.fromValues(x,y,z);
    }


    /**
     * Cette méthode supprime les ressources allouées
     */
    public void destroy()
    {
        m_Mesh.destroy();
    }
}

File: Shapes/Cone/WebGL/Cone.js
// Définition de la classe Cone

// superclasses et classes nécessaires
Requires("libs/Mesh/Mesh.js");
Requires("libs/Mesh/MeshModuleTopology.js");


// Définition de la classe Cone
class Cone
{
    /**
     * constructeur
     * @param triangle_count : nombre de secteurs (4 pour une pyramide à base carrée, bcp plus pour un cône)
     */
    constructor(triangle_count)
    {
        // créer le maillage : une grille rectangulaire refermée sur elle-même
        this.m_Mesh = new Mesh("Cone");
        let topology = new MeshModuleTopology(this.m_Mesh);
        let num0 = topology.addRectangularSurface(triangle_count, 2, "cone %d-%d", true, false);

        // définir les vertices du cône
        for (let is=0; is<triangle_count; is++) {
            // vertex sur la pointe
            let s0 = this.m_Mesh.getVertex(is + 0*triangle_count + num0);
            s0.setCoord(this.getCoord((is+0.5) / triangle_count, 0.0));

            // vertex sur le bord
            let s1 = this.m_Mesh.getVertex(is + 1*triangle_count + num0);
            s1.setCoord(this.getCoord((is+0.0) / triangle_count, 1.0));
        }

        // calculer les normales
        this.m_Mesh.computeNormals();
    }


    /**
     * fournit le maillage de cet objet
     * @return maillage
     */
    getMesh()
    {
        return this.m_Mesh;
    }


    /**
     * Cette méthode retourne les coordonnées 3D d'un point identifié par (a,b)
     * @param a : paramètre variant entre 0 et 1 en tournant autour du centre
     * @param b : rayon du point, distance à l'axe Y
     * @return vec3 contenant les coordonnées
     */
    getCoord(a,b)
    {
        // angle de ce point
        let angle = a * 2.0*Math.PI;

        // conversion en coordonnées cylindriques
        let x = b * Math.sin(angle);
        let y = 1.0 - b;
        let z = b * Math.cos(angle);
        return vec3.fromValues(x,y,z);
    }


    /**
     * Cette méthode supprime les ressources allouées
     */
    destroy()
    {
        this.m_Mesh.destroy();
    }
}
